packages/block-editor/src/components/resizable-alignment-controls/index.js

```diff
@@ -0,0 +1,213 @@
+	const [ nearestZone, setNearestZone ] = useState();
+	const zones = useBlockAlignmentZoneContext();
+
+	useEffect( () => {
+		if ( mousePosition ) {
+			const newNearestZone = throttledDetectNearestZone(
+				mousePosition,
+				zones
+			);
+			if ( newNearestZone !== nearestZone ) {
+				setNearestZone( newNearestZone );
+			}
+		}
+	}, [ mousePosition, zones ] );
+
+	const snapCoordinates = useMemo(
+		() => getSnapCoordinates( align, zones ),
+		[ align, zones, zones.size ]
+	);
+
+	let showRightHandle = false;
+	let showLeftHandle = false;
+
+	const rootClientId = useSelect(
+		( select ) =>
+			select( blockEditorStore ).getBlockRootClientId( clientId ),
+		[ clientId ]
+	);
+
+	/* eslint-disable no-lonely-if */
+	// See https://github.com/WordPress/gutenberg/issues/7584.
+	if ( align === 'center' ) {
+		// When the image is centered, show both handles.
+		showRightHandle = true;
+		showLeftHandle = true;
+	} else if ( isRTL() ) {
+		// In RTL mode the image is on the right by default.
+		// Show the right handle and hide the left handle only when it is
+		// aligned left. Otherwise always show the left handle.
+		if ( align === 'left' ) {
+			showRightHandle = true;
+		} else {
+			showLeftHandle = true;
+		}
+	} else {
+		// Show the left handle and hide the right handle only when the
+		// image is aligned right. Otherwise always show the right handle.
+		if ( align === 'right' ) {
+			showLeftHandle = true;
+		} else {
+			showRightHandle = true;
+		}
+	}
+	/* eslint-enable no-lonely-if */
+
+	return (
+		<>
+			<AnimatePresence>
+				{ isAlignmentVisualizerVisible && (
+					<motion.div
+						initial={ { opacity: 0 } }
+						animate={ { opacity: 1 } }
+						exit={ { opacity: 0 } }
+						transition={ { duration: 0.15 } }
+					>
+						<BlockAlignmentVisualizer
+							layoutClientId={ rootClientId }
+							focusedClientId={ clientId }
+							allowedAlignments={ allowedAlignments }
+							highlightedZone={ nearestZone }
+						/>
+					</motion.div>
+				) }
+			</AnimatePresence>
+			<ResizableBox
+				size={ size }
+				showHandle={ showHandle }
+				minWidth={ minWidth }
+				maxWidth={ maxWidth }
+				minHeight={ minHeight }
+				maxHeight={ maxHeight }
+				lockAspectRatio
+				enable={ {
+					top: false,
+					right: showRightHandle,
+					bottom: true,
+					left: showLeftHandle,
+				} }
+				onResizeStart={ ( ...resizeArgs ) => {
+					onResizeStart( ...resizeArgs );
+					const [ , resizeDirection ] = resizeArgs;
+					// TODO - reconsider supported alignments.
+					// Wide and Full currently don't show drag handles, but could do.
+					// Left and Right alignments could also work, but are trickier to implement.
+					if (
+						[ 'none', 'center' ].includes( align ) &&
+						( resizeDirection === 'right' ||
+							resizeDirection === 'left' )
+					) {
+						setIsAlignmentVisualizerVisible( true );
+					}
+				} }
+				onResize={ ( event ) => {
+					setMousePosition( { x: event.clientX, y: event.clientY } );
```

Comment:
This is causing this component to re-render every time the mouse position changes. And then re-render again in the `useEffect` because it's always a new object.

I think a more performant way to do this is to convert the `useEffect` to an event function, and only call `setNearestZone` when it needs to update.

```suggestion
					setNearestZone( throttledDetectNearestZone(
						{ x: event.clientX, y: event.clientY },
						zones
					) );
```

We also don't need to do the check between current and new values because `setState` will bail out by React if the value is the same.

Reply:
Sounds like a better option, thanks for the suggestion 👍